lib/aws-record/record/batch_writer.rb

```diff
@@ -0,0 +1,208 @@
+        validate_items
+        retry_items = send_chunks(opts)
+        responses = retry_failed_requests(retry_items, opts)
+        analyze_result(responses)
+      rescue RuntimeError => e
+        capture_validation_error(e)
+        raise e
+      end
+
+      # Perform the save! function but capture the error that is raised on
+      # unsuccessful save.
+      def save(opts = {})
+        save!(opts)
+        true
+      rescue RuntimeError
+        false
+      end
+
+      def error?
+        @state == ERROR_SEND_STATE
+      end
+
+      # Get a hash of the responses that didn't work along with their
+      # unprocessed items. Mimics ActiveRecord design.
+      def errors
+        @errors.map do |err|
+          {
+            error: err.error,
+            data: err.data,
+            unprocesssed_items: err.unprocessed_items
+          }
+        end
+      end
+
+      def pending?
+        @state == PENDING_STATE
+      end
+
+      def unprocessed_items
+        @errors.map(&:unprocessed_items)
+      end
+
+      def valid?
+        @state == SUCCESSFUL_SEND_STATE
+      end
+
+      private
+      def alert_errors(responses)
+        @errors += responses.reject(&:successful?)
+        @state = ERROR_SEND_STATE
+        raise Errors::SubmissionError
+      end
+
+      def analyze_result(responses)
+        alert_errors(responses) if failures?(responses)
+        @state = SUCCESSFUL_SEND_STATE
+      end
+
+      def batch_write(item_requests_array, opts)
+        @client.batch_write_item(
+          formatted_request(item_requests_array, opts)
+        )
+      end
+
+      def batch_write_items(item_requests_array, opts)
+        response = batch_write(item_requests_array, opts)
+        if response.successful?
+          response.unprocessed_items[@model.table_name] || []
+        else
+          item_requests_array
+        end
+      end
+
+      def capture_validation_error(e)
+        @state = ERROR_SEND_STATE
+        error = Struct.new(:error, :data, :unprocessed_items)
+        @errors << error.new(e.class, e, @items)
+      end
+
+      def failures?(responses)
+        responses.any? do |response|
+          !response.successful? || response.unprocessed_items.empty?
+        end
+      end
+
+      def formatted_request(item_requests_array, opts)
+        {
+          request_items: {
+            @model.table_name => item_requests_array
+          }
+        }.merge(opts)
+      end
+
+      def request_item(item)
+        {
+          put_request: {
+            item: item.build_item_for_save
+          }
+        }
+      end
+
+      def retry_failed_requests(item_requests_array, opts)
+        responses = []
+        item_requests_array.each_slice(ITEM_REQUEST_LIMIT) do |items|
+          responses << retry_chunk(items, opts)
+        end
+        responses
+      end
+
+      def retry_chunk(item_requests_array, opts)
+        response = nil
+        (0..4).each do |i|
+          sleep i**2
```

Comment:
I agree, it is too long. And i also agree that this isn't the best place to put this. I have refactored.